### DISCUSSION TITLE
My energy usage appears to be calculating incorrectly

### DISCUSSION THREAD

**Discussion by hkunzig** [teslamate-org/teslamate#1616] [Q&A]
I'm having an issue where recently my kw/m and total kwh numbers have jumped, retroactively.  I think I've tracked it to the efficiency column in the cars table.  I have a model Y standard with what appears to be a 50kw battery.  One of the reasons I'm sure it's calculating wrong is on the trips page, if I use a long time period, the kwh used is signnificantly more than th kwh added.  E.g. 800kwh added, 1300kwh used.
I've compared the wh/m displayed by the car and the wh/m computed, and it's off significantly.
The efficiency value computed in the cars table is 0.20915.  I've estimated the value to be 0.1276.  When I use that value, the wh/m in the car and displayed in teslamate match.
In log.ex I see the recalculate_efficiency procedure, and I tried to duplicate the query to try and find my problem.  I assume it's bad charge data, but I can't find it.   I also ran this query, which I derived that function:
select avg( charge_energy_added / (end_ideal_range_km - start_ideal_range_km)) from charging_processes
  where end_ideal_range_km>start_ideal_range_km
       and duration_min > 30 
The result is 0.13516 which is close to my estimate.

I also ran this query which didn't produce any extreme values:
select id,start_date,charge_energy_added / (end_ideal_range_km - start_ideal_range_km) from charging_processes
 where end_ideal_range_km>start_ideal_range_km
   and duration_min > 10 order by start_date;

Strangely though, there are 8 rows in the table on 5/7, that produce the same efficiency value (0.20915).  
The log said "Derived efficiency factor: 209.15 Wh/km (8x confirmed)"

So, I deleted those rows.  After the next charge, the efficiency factor was recomputed at 0.1361.
log: "Derived efficiency factor: 136.1 Wh/km (24x confirmed)"

Any thoughts on what's happening?  An easier way to fix it or detect it?  Should the recalculate_effieciency procedure maybe only include rows where energy_added>1.0 vs 0.0?

**Comment by adriankumpf** [comment]
Good analysis. Such measurement errors should normally be prevented by `duration_min > 10`. And since this calculation takes place after each charging session, this would have fixed itself after a while, I guess.  But I see no reason not to change the where condition to `energy_added>1.0`.
